tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+                if (ch == LF) {
+                    has_crlf = true;
+                } else {
+                    has_cr = true;
+                }
+            } else {
+                if (ch == LF) {
+                    has_lf = true;
+                }
+            }
+
+            if (ch == '\t') {
+                ++tab_characters;
+            } else if (ch == 0xEF || ch == 0xBB || ch == 0xBF) {
+                has_utf8_bom = true;
+            } else if (ch != CR && ch != LF && !(ch >= 0x20 && ch <= 0x7E)) {
+                ++disallowed_characters;
+                constexpr size_t MaxDisallowedCharacters = 10;
+                if (disallowed_characters <= MaxDisallowedCharacters) {
+                    fprintf(stderr, "Validation failed: %s contains disallowed character 0x%02X.\n", filepath.c_str(),
+                        static_cast<unsigned int>(ch));
+                }
+            }
+
+            if ((prev == ' ' || prev == '\t') && (ch == CR || ch == LF)) {
+                ++trailing_whitespace_lines;
+            }
+
+            (void) exchange(previous3, exchange(previous2, exchange(prev, ch)));
+        }
+    }
+
+    if (prev == CR) { // file ends with CR
+        has_cr = true;
+    }
+
+    if (has_cr) {
+        fprintf(stderr, "Validation failed: %s contains CR line endings (possibly damaged CRLF).\n", filepath.c_str());
+    } else if (has_lf && has_crlf) {
+        fprintf(stderr, "Validation failed: %s contains mixed line endings (both LF and CRLF).\n", filepath.c_str());
+    } else if (has_lf) {
+        fprintf(stderr, "Validation failed: %s contains LF line endings.", filepath.c_str());
+
+        if (prev != LF) {
+            fprintf(stderr, " Also, it doesn't end with a newline.\n");
+        } else if (previous2 == LF) {
+            fprintf(stderr, " Also, it ends with multiple newlines.\n");
+        } else {
+            fprintf(stderr, "\n");
+        }
+    } else if (has_crlf) {
+        if (previous2 != CR || prev != LF) {
+            fprintf(stderr, "Validation failed: %s doesn't end with a newline.\n", filepath.c_str());
+        } else if (previous3 == LF) {
+            fprintf(stderr, "Validation failed: %s ends with multiple newlines.\n", filepath.c_str());
+        }
+    } else {
+        fprintf(stderr, "Validation failed: %s doesn't contain any newlines.\n", filepath.c_str());
+    }
+
+    if (has_utf8_bom) {
+        fprintf(stderr, "Validation failed: %s contains UTF-8 BOM characters.\n", filepath.c_str());
+    }
+
+    if (tab_characters != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu tab characters.\n", filepath.c_str(), tab_characters);
+    }
+
+    if (trailing_whitespace_lines != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu lines with trailing whitespace.\n", filepath.c_str(),
+            trailing_whitespace_lines);
+    }
+}
+
+int main() {
+    static constexpr array skipped_directories{
+        ".git"sv,

```

Comment:
ideally this would just respect gitignore using `git check-ignore` or `git ls-files`

Reply:
For a widely reusable tool, I would agree. For this specific project, I believe that the effort needed to implement such centralization outweighs the effort of maintaining `.gitignore` and this skip list in parallel. Note that the skip list is needed in any event, since we may want to have directories in source control that contain binary files that shouldn't be processed by the validator.